File: backend/web/api/sync.py
import logging

from quart import current_app, jsonify

from .common import _get_provider_label

logger = logging.getLogger(__name__)


async def get_sync_status():
    """获取同步状态"""
    try:
        plugin_config = current_app.config.get("PLUGIN_CONFIG", {})
        category_manager = plugin_config.get("category_manager")

        if not category_manager:
            raise ValueError("未找到类别管理器")

        logger.info("获取同步状态...")
        missing_in_config, deleted_categories = category_manager.get_sync_status()

        return jsonify(
            {
                "status": "ok",
                "missing_in_config": missing_in_config,
                "deleted_categories": deleted_categories,
                "differences": {
                    "missing_in_config": missing_in_config,
                    "deleted_categories": deleted_categories,
                },
            }
        )
    except Exception as e:
        logger.error(f"获取同步状态失败: {e}")
        return jsonify({"error": "获取同步状态失败"}), 500


async def sync_config():
    """同步配置与文件夹结构的 API 端点"""
    try:
        plugin_config = current_app.config.get("PLUGIN_CONFIG", {})
        category_manager = plugin_config.get("category_manager")

        if not category_manager:
            raise ValueError("未找到类别管理器")

        logger.info("开始同步配置...")
        if category_manager.sync_with_filesystem():
            logger.info("配置同步成功")
            return jsonify({"message": "配置同步成功"}), 200
        else:
            logger.warning("配置同步失败")
            return jsonify({"message": "配置同步失败"}), 500
    except Exception as e:
        logger.error(f"配置同步失败: {e}")
        return jsonify({"message": f"配置同步失败: {str(e)}"}), 500


async def get_img_host_sync_status():
    """获取同步状态"""
    try:
        plugin_config = current_app.config.get("PLUGIN_CONFIG", {})
        img_sync = plugin_config.get("img_sync")
        if not img_sync:
            return jsonify({"error": "图床服务未配置"}), 400

        status = img_sync.check_status()
        status["upload_count"] = len(status.get("to_upload", []))
        status["download_count"] = len(status.get("to_download", []))
        status["provider_label"] = _get_provider_label(img_sync)
        return jsonify(status)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


async def sync_to_remote():
    """同步到云端"""
    try:
        plugin_config = current_app.config.get("PLUGIN_CONFIG", {})
        img_sync = plugin_config.get("img_sync")
        if not img_sync:
            return jsonify({"message": "图床服务未配置"}), 400

        img_sync.sync_process = img_sync._start_sync_process("upload")
        return jsonify({"success": True})
    except Exception as e:
        return jsonify({"message": str(e)}), 500


async def sync_from_remote():
    """从云端同步"""
    try:
        plugin_config = current_app.config.get("PLUGIN_CONFIG", {})
        img_sync = plugin_config.get("img_sync")
        if not img_sync:
            return jsonify({"message": "图床服务未配置"}), 400

        img_sync.sync_process = img_sync._start_sync_process("download")
        return jsonify({"success": True})
    except Exception as e:
        return jsonify({"message": str(e)}), 500


async def check_sync_process():
    """检查同步进程状态"""
    try:
        plugin_config = current_app.config.get("PLUGIN_CONFIG", {})
        img_sync = plugin_config.get("img_sync")
        if not img_sync or not img_sync.sync_process:
            return jsonify({"completed": True, "success": True})

        if not img_sync.sync_process.is_alive():
            success = img_sync.sync_process.exitcode == 0
            img_sync.sync_process = None
            return jsonify({"completed": True, "success": success})

        return jsonify({"completed": False})
    except Exception as e:
        return jsonify({"message": str(e)}), 500

File: pages/meme_manager/js/modules/sync.js
const { ref } = window.Vue;

export function useSync(showToast, fetchEmojis) {
  const syncChecking = ref(false);
  const syncStatus = ref({
    inSync: true,
    missingInConfig: [],
    deletedCategories: [],
  });

  const imgHostSyncing = ref(false);
  const imgHostStatus = ref({
    provider: "--",
    remoteImageCount: "--",
    remoteStorageSize: "--",
    uploadCount: 0,
    downloadCount: 0,
  });

  const formatBytes = (bytes) => {
    if (typeof bytes !== "number" || Number.isNaN(bytes) || bytes < 0) return "未知";
    if (bytes === 0) return "0 B";
    const units = ["B", "KB", "MB", "GB", "TB"];
    let value = bytes;
    let unitIndex = 0;
    while (value >= 1024 && unitIndex < units.length - 1) {
      value /= 1024;
      unitIndex++;
    }
    return `${value.toFixed(1)} ${units[unitIndex]}`;
  };

  const checkSyncStatus = async (showAlert = true) => {
    syncChecking.value = true;
    try {
      const res = await fetch("/api/sync/status");
      if (!res.ok) throw new Error("获取状态失败");
      const data = await res.json();

      if (data.status === "error") throw new Error(data.message);

      const diffs = data.differences || data;
      syncStatus.value.missingInConfig = diffs.missing_in_config || [];
      syncStatus.value.deletedCategories = diffs.deleted_categories || [];
      syncStatus.value.inSync =
        syncStatus.value.missingInConfig.length === 0 && syncStatus.value.deletedCategories.length === 0;

      if (showAlert) {
        showToast("配置同步状态检查完毕。", "success", "刷新成功");
      }
    } catch (e) {
      showToast(e.message, "error", "同步检查失败");
    } finally {
      syncChecking.value = false;
    }
  };

  const syncConfig = async () => {
    try {
      const res = await fetch("/api/sync/config", { method: "POST" });
      if (!res.ok) throw new Error("同步失败");
      showToast("已成功将磁盘文件夹同步至系统配置", "success", "配置同步完成");
      await fetchEmojis();
      await checkSyncStatus(false);
    } catch (e) {
      showToast(e.message, "error", "同步失败");
    }
  };

  const removeFromConfig = async (category) => {
    try {
      const res = await fetch("/api/category/remove", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ category }),
      });
      if (!res.ok) throw new Error("移除配置失败");
      showToast(`已从配置中移除标签 「${category}」`, "success", "移除成功");
      await fetchEmojis();
      await checkSyncStatus(false);
    } catch (e) {
      showToast(e.message, "error", "删除配置失败");
    }
  };

  const checkImgHostSyncStatus = async (showAlert = true) => {
    try {
      const res = await fetch("/api/img_host/sync/status");
      if (!res.ok) throw new Error("获取图床状态失败");
      const data = await res.json();

      const uploadCount = data.upload_count ?? data.to_upload?.length ?? 0;
      const downloadCount = data.download_count ?? data.to_download?.length ?? 0;
      const remoteImageCount = data.remote_image_count ?? data.remote_count ?? data.remote_images?.length ?? 0;
      
      let remoteStorageText = "未知";
      if (typeof data.remote_total_bytes === "number") {
        remoteStorageText = formatBytes(data.remote_total_bytes);
      } else if (typeof data.remote_total_bytes_estimated === "number") {
        remoteStorageText = `${formatBytes(data.remote_total_bytes_estimated)}（估算）`;
      }

      imgHostStatus.value.provider = data.provider_label || "未知图床";
      imgHostStatus.value.remoteImageCount = remoteImageCount;
      imgHostStatus.value.remoteStorageSize = remoteStorageText;
      imgHostStatus.value.uploadCount = uploadCount;
      imgHostStatus.value.downloadCount = downloadCount;

      if (showAlert) {
        showToast(
          `${data.provider_label || "图床"}已刷新。云端：${remoteImageCount} 张，待上传：${uploadCount} 个。`,
          "success",
          "刷新成功"
        );
      }
    } catch (e) {
      showToast(e.message, "error", "图床同步检查失败");
    }
  };

  const syncToRemote = async () => {
    imgHostSyncing.value = true;
    showToast("正在将待上传文件同步至云端图床...", "info", "上传同步中", 5000);
    try {
      const res = await fetch("/api/img_host/sync/upload", { method: "POST" });
      if (!res.ok) throw new Error("上传同步失败");
      showToast("本地表情包成功全量同步至云端图床", "success", "同步成功");
      await checkImgHostSyncStatus(false);
    } catch (e) {
      showToast(e.message, "error", "同步失败");
    } finally {
      imgHostSyncing.value = false;
    }
  };

  const syncFromRemote = async () => {
    imgHostSyncing.value = true;
    showToast("正在从云端下载表情包...", "info", "下载同步中", 5000);
    try {
      const res = await fetch("/api/img_host/sync/download", { method: "POST" });
      if (!res.ok) throw new Error("下载同步失败");
      showToast("云端图床成功全量拉取同步至本地", "success", "同步成功");
      await fetchEmojis();
      await checkImgHostSyncStatus(false);
    } catch (e) {
      showToast(e.message, "error", "同步失败");
    } finally {
      imgHostSyncing.value = false;
    }
  };

  return {
    syncChecking,
    syncStatus,
    imgHostSyncing,
    imgHostStatus,
    formatBytes,
    checkSyncStatus,
    syncConfig,
    removeFromConfig,
    checkImgHostSyncStatus,
    syncToRemote,
    syncFromRemote,
  };
}
